ud/uct_flush_test.put_bcopy_flush_ep_no_comp hang

http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2132/console
```
12:21:29 [       OK ] ud/uct_flush_test.put_bcopy_flush_ep_no_comp/1 (98 ms)
12:21:29 [ RUN      ] ud/uct_flush_test.am_pending_flush_nb/1
13:26:29 make: *** [test] Terminated
```

http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2141/console
```
15:13:32 [ RUN      ] ud_mlx5/uct_flush_test.am_pending_flush_nb/0
16:25:13 make: *** [test] Terminated
16:25:13 Build step 'Execute shell' marked build as failure
```

--- a/src/ucs/datastruct/arbiter.c
+++ b/src/ucs/datastruct/arbiter.c
@@ -98,7 +98,7 @@ void ucs_arbiter_group_schedule_nonempty(ucs_arbiter_t *arbiter,
     head = tail->next;
 
     if (head == NULL) {
-        /* it means that 1 element group is 
+        /* it means that 1 element group is
          * scheduled during dispatch.
          * Restore next pointer.
          */
@@ -164,7 +164,7 @@ void ucs_arbiter_dispatch_nonempty(ucs_arbiter_t *arbiter, unsigned per_group,
             ++group_dispatch_count;
 
             if (result == UCS_ARBITER_CB_RESULT_REMOVE_ELEM) {
-               if (elem == last_elem) {
+                 if (elem == last_elem) {
                     /* Only element */
                     group->tail = NULL; /* Group is empty now */
                     if (group_head == prev_group) {

--- a/src/ucs/datastruct/arbiter.h
+++ b/src/ucs/datastruct/arbiter.h
@@ -168,23 +168,6 @@ void ucs_arbiter_group_push_elem_always(ucs_arbiter_group_t *group,
                                         ucs_arbiter_elem_t *elem);
 
 /**
- * Add a new work element to a group if it is not already there
- *
- * @param [in]  group    Group to add the element to.
- * @param [in]  elem     Work element to add.
- */
-static inline void 
-ucs_arbiter_group_push_elem(ucs_arbiter_group_t *group,
-                            ucs_arbiter_elem_t *elem)
-{
-    if (elem->next != NULL) {
-        return;
-    }
-    ucs_arbiter_group_push_elem_always(group, elem);
-}
-
-
-/**
  * Remove all elements from a group, and call the callback for each of them.
  * Callback return value is ignored.
  *
@@ -195,11 +178,6 @@ ucs_arbiter_group_push_elem(ucs_arbiter_group_t *group,
  */
 void ucs_arbiter_group_purge(ucs_arbiter_t *arbiter, ucs_arbiter_group_t *group,
                              ucs_arbiter_callback_t cb, void *cb_arg);
-
-static inline int ucs_arbiter_group_is_empty(ucs_arbiter_group_t *group)
-{
-    return group->tail == NULL;
-}
 
 void ucs_arbiter_dump(ucs_arbiter_t *arbiter, FILE *stream);
 
@@ -215,6 +193,27 @@ void ucs_arbiter_dispatch_nonempty(ucs_arbiter_t *arbiter, unsigned per_group,
 /* Internal function */
 void ucs_arbiter_group_head_desched(ucs_arbiter_t *arbiter,
                                     ucs_arbiter_elem_t *head);
+
+
+/**
+ * Return true if arbiter has no groups scheduled
+ *
+ * @param [in]  arbiter    Arbiter object to dispatch work on.
+ */
+static inline int ucs_arbiter_is_empty(ucs_arbiter_t *arbiter)
+{
+    return arbiter->current == NULL;
+}
+
+
+/**
+ * @return whether if the group does not have any queued elements.
+ */
+static inline int ucs_arbiter_group_is_empty(ucs_arbiter_group_t *group)
+{
+    return group->tail == NULL;
+}
+
 
 /**
  * Schedule a group for arbitration. If the group is already there, the operation
@@ -252,15 +251,31 @@ static inline void ucs_arbiter_group_desched(ucs_arbiter_t *arbiter,
 }
 
 /**
- * Return true if arbiter has no groups scheduled
- *
- * @param [in]  arbiter    Arbiter object to dispatch work on.
+ * @return whether the group does not have any queued elements.
  */
-static inline int
-ucs_arbiter_is_empty(ucs_arbiter_t *arbiter)
+static inline int ucs_arbiter_elem_is_scheduled(ucs_arbiter_elem_t *elem)
 {
-    return arbiter->current == NULL;
+    return elem->next != NULL;
 }
+
+
+/**
+ * Add a new work element to a group if it is not already there
+ *
+ * @param [in]  group    Group to add the element to.
+ * @param [in]  elem     Work element to add.
+ */
+static inline void
+ucs_arbiter_group_push_elem(ucs_arbiter_group_t *group,
+                            ucs_arbiter_elem_t *elem)
+{
+    if (ucs_arbiter_elem_is_scheduled(elem)) {
+        return;
+    }
+
+    ucs_arbiter_group_push_elem_always(group, elem);
+}
+
 
 /**
  * Dispatch work elements in the arbiter. For every group, up to per_group work

--- a/src/uct/ib/ud/base/ud_ep.c
+++ b/src/uct/ib/ud/base/ud_ep.c
@@ -677,7 +677,7 @@ ucs_status_t uct_ud_ep_flush_nolock(uct_ud_iface_t *iface, uct_ud_ep_t *ep,
              * Also, prevent from sending more control messages like this after
              * first time by turning on the flag on the last skb.
              */
-            uct_ud_ep_ctl_op_add(iface, ep, UCT_UD_EP_OP_ACK_REQ);
+            uct_ud_ep_ctl_op_add_safe(iface, ep, UCT_UD_SEND_SKB_FLAG_ACK_REQ);
             skb->flags |= UCT_UD_SEND_SKB_FLAG_ACK_REQ;
         }
 
@@ -1002,6 +1002,7 @@ ucs_status_t uct_ud_ep_pending_add(uct_ep_h ep_h, uct_pending_req_t *req)
     ucs_arbiter_group_push_elem(&ep->tx.pending.group, 
                                 (ucs_arbiter_elem_t *)req->priv);
     ucs_arbiter_group_schedule(&iface->tx.pending_q, &ep->tx.pending.group);
+
     iface->tx.pending_q_len++;
     uct_ud_leave(iface);
     return UCS_OK;

--- a/src/uct/ib/ud/base/ud_inl.h
+++ b/src/uct/ib/ud/base/ud_inl.h
@@ -1,3 +1,17 @@
+/**
+ * Copyright (C) Mellanox Technologies Ltd. 2001-2014.  ALL RIGHTS RESERVED.
+ *
+ * See file LICENSE for terms.
+ */
+
+static UCS_F_ALWAYS_INLINE void
+uct_ud_ep_ctl_op_schedule(uct_ud_iface_t *iface, uct_ud_ep_t *ep)
+{
+    ucs_assert(!iface->tx.in_pending);
+    ucs_arbiter_group_push_elem(&ep->tx.pending.group,
+                                &ep->tx.pending.elem);
+    ucs_arbiter_group_schedule(&iface->tx.pending_q, &ep->tx.pending.group);
+}
 
 /**
  * schedule control operation.
@@ -5,10 +19,17 @@
 static UCS_F_ALWAYS_INLINE void
 uct_ud_ep_ctl_op_add(uct_ud_iface_t *iface, uct_ud_ep_t *ep, int op)
 {
-    ucs_arbiter_group_push_elem(&ep->tx.pending.group, 
-                                &ep->tx.pending.elem);
     ep->tx.pending.ops |= op;
-    ucs_arbiter_group_schedule(&iface->tx.pending_q, &ep->tx.pending.group);
+    uct_ud_ep_ctl_op_schedule(iface, ep);
+}
+
+static UCS_F_ALWAYS_INLINE void
+uct_ud_ep_ctl_op_add_safe(uct_ud_iface_t *iface, uct_ud_ep_t *ep, int op)
+{
+    ep->tx.pending.ops |= op;
+    if (!iface->tx.in_pending) {
+        uct_ud_ep_ctl_op_schedule(iface, ep);
+    }
 }
 
 /* 

